TACC/Core-CMS
# feat(tup-308): working form plugin (aka tup-230)

## Overview

Update and configure form plugin to work.

## Related

- [TUP-308](https://jira.tacc.utexas.edu/browse/TUP-308)
- requires https://github.com/TACC/tup-ui/pull/25
- mimicked by https://github.com/TACC/Core-CMS/pull/500 but on v3.6.0
- mimics https://github.com/TACC/Core-CMS/pull/501 (also from v3.6.0)

## Changes

- install `django-recaptcha`
- replace `djangocms-forms` with `djangocms-forms-maintained`
- isolate and document form plugin settings
- extend `c-button` styles (just a little)
- style form elements (generic and plugin-specific)
- load jQuery with AJAX (to support form)
- sample code to customize captcha
- add required url entry
- add Frontera colors\* for primary button

<sup>* I use Frontera colors only until we have standard form design. M.S. knows he has this task, and has screenshots to work know what all needs to be designed.</sup>

## Testing

Remote Test Pages:

- [Core (Dev)](https://dev.cep.tacc.utexas.edu/test/tup-230-form-plugin/?edit_off=true)
- [Frontera (Pre-Prod)](https://pprd.frontera-portal.tacc.utexas.edu/test/tup-230-form-plugin/) (or a [form with all fields](https://pprd.frontera-portal.tacc.utexas.edu/test/tup-320-form-plugin/))
- [APCD (Pre-Prod)](https://apcd-qa.tacc.utexas.edu/test/tup-230-form-plugin/)

Create, Edit, Use:

1. (Optional) Create form plugin instance. While editing form:
    - Complete _required_ fields (of form editor),
    but [choose save to database](https://user-images.githubusercontent.com/62723358/179061534-ed48b6e0-4a5d-4711-877d-729bfa496fbd.png), not send e-mail.[^1]
3. Submit form—see success message.[^2]
4. Submit form again—see success message.[^2]
5. Edit form plugin instance.
    - Add SPAM PROTECTION[^3]
6. Submit form—see success message.[^2]
7. In "Site administration", open "Form submissions".
8. Confirm your submissions are there.

Administration:

1. View submissions in admin "Form submissions".
2. Export CSV and JSON from admin "Form submissions".

[^1]: E-mail feature is delayed until I have bandwidth to research it and configure or fix it.
[^2]: The success message, for editors only, is at the top center of screen white on black (a `.cms_message`). This does not happen when a user submits, or while editor is seeing it as "View published".
[^3]: There should only be options "None" and "Recaptcha"; choose "Recaptcha".

## UI

<details open><summary>Forms on Servers To Which I Deployed</summary>

| local | core | frontera | apcd |
| - | - | - | - |
| ![local](https://user-images.githubusercontent.com/62723358/177679719-9aeb226a-3e74-44af-afe9-43b0d7ae2de0.jpeg) | ![core](https://user-images.githubusercontent.com/62723358/177679715-6223d744-c02c-464b-86a2-a5717d9cea4d.jpeg) | ![frontera](https://user-images.githubusercontent.com/62723358/177679724-c34e5067-b96b-486a-b735-1b58834cc53b.jpeg) | ![apcd](https://user-images.githubusercontent.com/62723358/177679726-bd96a23b-11b2-4b9e-bd66-f6e4a41f29dd.jpeg)


</details>

[TUP-275]: https://jira.tacc.utexas.edu/browse/TUP-275
